### PR TITLE
Add ST7735 TFT driver and switch SPI to full duplex

### PR DESCRIPTION
**Pull request by mmoskal**

### CODE REVIEW

File: source/driver-models/SPI.cpp [hunk 1]
```diff
@@ -24,30 +24,45 @@ DEALINGS IN THE SOFTWARE.
...
 }
+
+/**
+ * Writes and reads from the SPI bus concurrently. Finally, calls doneHandler (possibly in IRQ
+ * context).
+ *
+ * Either buffer can be NULL.
+ */
+int SPI::startTransfer(const uint8_t *txBuffer, uint32_t txSize, uint8_t *rxBuffer, uint32_t rxSize,
+                       PVoidCallback doneHandler, void *arg)
```
Comment (jamesadevine) [inline source/driver-models/SPI.cpp]: Why not a combination of `fiber_wait_for_event` and Event notify, as per MicroBitDisplay?

File: source/drivers/ST7735.cpp [hunk 2]
```diff
@@ -0,0 +1,352 @@
...
+    work->srcLeft = ((width + 1) >> 1) * height;
+    work->x = 0;
+
+    sendColorsStep(this);
+
+    return DEVICE_OK;
+}
+
+// we don't modify *buf, but it cannot be in flash, so no const as a hint
+void ST7735::sendCmd(uint8_t *buf, int len)
```
Comment (jamesadevine) [inline source/drivers/ST7735.cpp]: It seems like `sendCmd`, `sendBytes`, and `sendWords` should actually be the interfaces for the SPI abstraction?

File: source/driver-models/SPI.cpp (hunk 1 above)
Comment (mmoskal) [reply source/driver-models/SPI.cpp]: I believe the performance of message queue (even in immediate mode) will not be enough. On NRF52 you can only send 255 bytes at a time over SPI, which at 8Mhz is around 2000 cycles. After each 255 bytes I need this callback to prepare more data.

Walking over listener queue looks like at least 20 cycles per entry and maybe 200 for the actual invocation of my handler. With 30 active handlers this is at least 800 cycles, plus some overheads. At 8MHz I refresh the screen continuously (for 30fps), so I would be burning 50% of the CPU for handler invocation. This is before we even talk about context switch (so events in immediate mode).

Honestly, even if we reduce that to say 25% it still seems too much for something that is just an overhead in this scenario.

Comment (jamesadevine) [reply source/driver-models/SPI.cpp]: Makes sense, definitely needs to be as performant as possible :smile:

**Pull request merged**